src/gui/qgsfieldmappingmodel.cpp

```diff
@@ -0,0 +1,434 @@
+  }
+  beginMoveRows( QModelIndex( ), row, row, QModelIndex(), row + 2 );
+#if QT_VERSION < QT_VERSION_CHECK(5, 13, 0)
+  mMapping.swap( row, row + 1 );
+#else
+  mMapping.swapItemsAt( row, row + 1 );
+#endif
+  endMoveRows();
+  return true;
+}
+
+QString QgsFieldMappingModel::bestMatchforField( const QgsFieldMappingModel::Field &f, QStringList &excludedFieldNames )
+{
+  // Search for fields in the source
+  // 1. match by name
+  for ( const auto &sf : qgis::as_const( mSourceFields ) )
+  {
+    if ( sf.name() == f.field.name() )
+    {
+      return QgsExpression::quotedColumnRef( sf.name() );
+    }
+  }
+  // 2. match by type
+  for ( const auto &sf : qgis::as_const( mSourceFields ) )
+  {
+    if ( excludedFieldNames.contains( sf.name() ) || sf.type() != f.field.type() )
+      continue;
+    excludedFieldNames.push_back( sf.name() );
+    return QgsExpression::quotedColumnRef( sf.name() );
+  }
+  return QString();
+}
+
+void QgsFieldMappingModel::setSourceFields( const QgsFields &sourceFields )
+{
+  mSourceFields = sourceFields;
+  QStringList usedFields;
+  beginResetModel();
+  for ( const Field &f : qgis::as_const( mMapping ) )
+  {
+    if ( f.expression.isField() )
+    {
+      usedFields.push_back( f.expression.expression().mid( 1, f.expression.expression().length() - 2 ) );
+    }
+  }
+  // not const on purpose
+  for ( Field &f : mMapping )
+  {
+    if ( f.expression.expression().isEmpty() )
+    {
+      const QString expression { bestMatchforField( f, usedFields ) };
+      if ( ! expression.isEmpty() )
+        f.expression = expression;
+    }
+  }
+  endResetModel();
+}
+
+void QgsFieldMappingModel::setDestinationFields( const QgsFields &destinationFields,
+    const QMap<QString, QgsExpression> &expressions )
+{
+  beginResetModel();
+  mMapping.clear();
+  // Prepare the model data
+  QStringList usedFields;
+  for ( const auto &df : qgis::as_const( destinationFields ) )
+  {
+    Field f;
+    f.field = df;
+    f.originalName = df.name();
+    if ( expressions.contains( f.field.name() ) )
+    {
+      f.expression = expressions.value( f.field.name() );
+      // if it's source field
+      if ( f.expression.isField() &&
+           mSourceFields.names().contains( f.expression.referencedColumns().toList().first() ) )
+      {
+        usedFields.push_back( f.expression.referencedColumns().toList().first() );
+      }
+    }
+    else
+    {
+      const QString expression { bestMatchforField( f, usedFields ) };
+      if ( ! expression.isEmpty() )
+        f.expression = expression;
+    }
+    mMapping.push_back( f );
+  }
+  endResetModel();
+}
+
+bool QgsFieldMappingModel::destinationEditable() const
+{
+  return mDestinationEditable;
+}
+
+void QgsFieldMappingModel::setDestinationEditable( bool destinationEditable )
+{
+  mDestinationEditable = destinationEditable;
+}
+
+const QMap<QVariant::Type, QString> QgsFieldMappingModel::dataTypes() const
+{
+  static const QMap<QVariant::Type, QString> sDataTypes
```

Comment:
maybe add a comment here (and in the memory provider) that the two lists needs to be synchronized. (Or better - move it to a static member in QgsVectorDataProvider? Something like `allKnownFieldTypes` ?

Reply:
well, they are not the same list: in the memory provider we have multiple supported entries for the same QVariant::Type, I simplified the list here by providing a single entry for each of the duplicated QVariant::Type basic types.